test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (c) 2024 Red Hat and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2;
+
+import compiler.lib.ir_framework.*;
+import compiler.lib.ir_framework.Test;
+
+/*
+ * @test
+ * @bug 8325495
+ * @summary C2 should optimize for series of Add of unique value. e.g., a + a + ... + a => a*n
+ * @library /test/lib /
+ * @run driver compiler.c2.TestSerialAdditions
+ */
+public class TestSerialAdditions {
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @DontInline
+    private static void verifyResult(int base, int factor, int observed) {
+        int expected = base * factor; // compute expected result here while making sure not inlined in callers
+        if (expected != observed) {
+            throw new AssertionError("Expected " + expected + " but got " + observed);
+        }
+    }
+
+    @DontInline
+    private static void verifyResult(long base, long factor, long observed) {
+        long expected = base * factor;
+        if (expected != observed) {
+            throw new AssertionError("Expected " + expected + " but got " + observed);
+        }
+    }
+
+    // ----- integer tests -----
+    @Test
+    @Arguments(values = {Argument.RANDOM_EACH})
+    @IR(counts = { IRNode.ADD_I, "1" })
+    @IR(failOn = {IRNode.LSHIFT_I})
+    private static void addTo2(int a) {
+        int sum = a + a; // Simple additions like a + a should be kept as-is
+        verifyResult(a, 2, sum);
```

Comment:
Generally, we should move all verification code out of the `@Test` methods to avoid side effects and worrying about whether the result checking is now compiled or not (we must ensure that the result checking code is interpreted to catch wrong executions with miscompiled code).

I suggest the following (not tested):

Introduce a `@Run` method, which is never compiled, for your `@Test` methods. You can still call methods from there but then you should ensure that they are not compiled either with `@DontCompile`:
```
    static final Random RANDOM = Utils.getRandomInstance();

    ...

    @DontCompile
    private static void verifyResult(int base, int factor, int observed) { ... }

    ...

    @Test
    @IR(counts = {IRNode.ADD_I, "1", IRNode.LSHIFT_I, "1"})
    private static int addTo3(int a) {
        return a + a + a; // a*3 => (a<<1) + a
    }

    @Run(test = "addTo3") 
    void runAddTo3() {
        int a = RANDOM.nextInt();
        int result = addTo3(a);
        verifyResult(a, 3, result);
    }
```
Since the tests are all very similar and require the same setup and verification, you could even go a step further and provide a single shared `@Run` method which is possible:
```
    @Test
    @IR(counts = {IRNode.ADD_I, "1", IRNode.LSHIFT_I, "1"})
    private static int addTo3(int a) {
        return a + a + a; // a*3 => (a<<1) + a
    }

    @Test
    @IR(failOn = IRNode.ADD_I)
    @IR(counts = {IRNode.LSHIFT_I, "1"})
    private static int addTo4(int a) {
        return a + a + a + a; // a*4 => a<<2
    }

    @Run(test = {"addTo3", "addTo4"}) // List all @Test methods here and make sure you call all of them below.
    void runTests() {
        int a = RANDOM.nextInt();
        verifyResult(a, 3, addTo3(a));
        verifyResult(a, 4, addTo4(a));
    }

```
This also allows you to run with some more edge case values like `a == 0` or `a == min_int` etc. which gives us even some more confidence.

Reply:
Thanks for the idea. Converted to custom `@Run` methods and test with `a = 0, 1, min, max, rand`